storage/src/nodestore.rs

```diff
@@ -691,18 +642,99 @@ struct Version {
+        op: semver::Op::GreaterEq,
+        major: 0,
+        minor: Some(0),
+        patch: Some(4),
+        pre: semver::Prerelease::EMPTY,
+    };
+
+    /// Validates that the version identifier is valid and compatible with the current
+    /// build of firewood.
+    ///
+    /// # Errors
+    ///
+    /// - If the token contains invalid utf-8 bytes (nul is allowed).
+    /// - If the token does not start with "firewood ".
+    /// - If the version is not parsable by [`semver::Version`].
+    /// - If the version is not compatible with the current build of firewood.
+    ///   - Currently, the minimum required version is 0.0.4.
+    fn validate(&self) -> Result<(), Error> {
+        let version = std::str::from_utf8(&self.bytes).map_err(|e| {
+            Error::new(
+                ErrorKind::InvalidData,
+                format!("Invalid database version: invalid utf-8: {e}"),
+            )
+        })?;
+
+        // strip trailling nuls as they're only for padding
+        let version = version.trim_end_matches('\0');
 
-    /// construct a [Version] header from the firewood version
+        // strip magic prefix or error
+        let version = version.strip_prefix("firewood ").ok_or_else(|| {
+            Error::new(
+                ErrorKind::InvalidData,
+                format!(
+                    "Invalid database version: does not start with magic 'firewood ': {version}",
+                ),
+            )
+        })?;
+
+        // Version strings from CARGO_PKG_VERSION are guaranteed to be parsable by
+        // semver (cargo uses the same library).
+        let version = semver::Version::parse(version).map_err(|e| {
+            Error::new(
+                ErrorKind::InvalidData,
+                format!("Invalid version string: {version}: {e}"),
+            )
+        })?;
+
+        // verify base compatibility version
+        if !Self::BASE_VERSION.matches(&version) {
+            return Err(Error::new(
+                ErrorKind::InvalidData,
+                format!(
+                    "Database was created with firewood version {version}; however, this build of firewood requires version {}",
+                    Self::BASE_VERSION,
+                ),
+            ));
+        }
+
+        trace!("Database version is valid: {version}");
+        Ok(())
+    }
+
+    /// Construct a [`Version`] instance for the current build of firewood.
     fn new() -> Self {
-        let mut version_bytes: [u8; Self::SIZE as usize] = Default::default();
-        let version = env!("CARGO_PKG_VERSION");
-        let _ = version_bytes
-            .as_mut_slice()
-            .write_all(format!("firewood {}", version).as_bytes());
-        Self {
-            bytes: version_bytes,
+        // Note that with this magic token of 9 bytes, we can store a version string of
+        // up to 7 bytes. If we always include the major, minor, and patch versions,
+        // then no more than two of three can be 2 digits long.
```

Comment:
what a mess. I wonder if we can shorten "firewood" to "fwd" going forward. Sounds like a different PR for someone else to address later though.